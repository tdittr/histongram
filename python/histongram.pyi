from typing import Iterator


class Histogram:
    def __init__(self): ...

    def add(self, key: str): ...

    def add_many(self, iter: Iterator[str]): ...

    def __getitem__(self, key: str) -> int: ...
