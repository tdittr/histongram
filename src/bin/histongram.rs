use std::fs::read_to_string;
use std::path::PathBuf;

use clap::Parser;
use color_eyre::Result;
use compact_str::CompactString;

use histongram::{Histogram, Ngrams};

#[derive(Parser, Debug, Clone)]
struct Args {
    file: PathBuf,

    #[clap(long)]
    print: bool,
}

fn main() -> Result<()> {
    color_eyre::install()?;
    let args = Args::parse();

    let data = read_to_string(args.file)?;

    let ngrams = Ngrams::new(1..=5).count(data.split_whitespace());

    if args.print {
        println!("{ngrams:?}");
    }

    Ok(())
}
