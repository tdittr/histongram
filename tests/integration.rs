use std::cmp::{min, Reverse};

use insta::assert_ron_snapshot;

use histongram::Histogram;

const APACHE: &str = include_str!("../LICENSE-APACHE");
const MIT: &str = include_str!("../LICENSE-MIT");

#[test]
fn license_words() {
    let mut a = Histogram::from_iter(APACHE.split_whitespace());
    let mut a_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut a_counts);
    assert_ron_snapshot!(a_counts);

    let m = Histogram::from_iter(MIT.split_whitespace());
    let mut m_counts = m.clone().sorted_occurrences();
    sort_also_by_key(&mut m_counts);
    assert_ron_snapshot!(m_counts);

    a.append(m);
    let mut combined_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut combined_counts);
    assert_ron_snapshot!(combined_counts);
}

#[test]
fn license_chars() {
    let mut a = Histogram::from_iter(APACHE.chars());
    let mut a_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut a_counts);
    assert_ron_snapshot!(a_counts);

    let m = Histogram::from_iter(MIT.chars());
    let mut m_counts = m.clone().sorted_occurrences();
    sort_also_by_key(&mut m_counts);
    assert_ron_snapshot!(m_counts);

    a.append(m);
    let mut combined_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut combined_counts);
    assert_ron_snapshot!(combined_counts);
}

fn sort_also_by_key<K: Ord + Copy>(counts: &mut [(K, usize)]) {
    // This table should be sorted by count. But items with the same number of occurrences are
    // in arbitrary order...
    let mut high = usize::MAX;
    for &(_, cnt) in counts.iter() {
        assert!(cnt <= high);
        high = min(high, cnt);
    }

    // Now sort also by key so the snapshot stays consistent
    counts.sort_unstable_by_key(|(key, cnt)| (Reverse(*cnt), *key));
}
