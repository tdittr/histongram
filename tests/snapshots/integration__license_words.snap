---
source: tests/integration.rs
expression: a_counts
---
[
  ("the", 97),
  ("or", 62),
  ("of", 61),
  ("and", 41),
  ("to", 39),
  ("any", 28),
  ("You", 23),
  ("in", 23),
  ("that", 22),
  ("Work", 20),
  ("a", 20),
  ("by", 20),
  ("for", 18),
  ("Derivative", 17),
  ("such", 17),
  ("this", 16),
  ("as", 15),
  ("shall", 15),
  ("Works", 12),
  ("License.", 11),
  ("not", 11),
  ("on", 11),
  ("with", 11),
  ("copyright", 10),
  ("mean", 10),
  ("License", 9),
  ("may", 9),
  ("under", 9),
  ("Contributor", 8),
  ("Licensor", 8),
  ("Source", 8),
  ("Your", 8),
  ("form", 8),
  ("is", 8),
  ("notices", 8),
  ("within", 8),
  ("conditions", 7),
  ("other", 7),
  ("terms", 7),
  ("License,", 6),
  ("an", 6),
  ("are", 6),
  ("file", 6),
  ("NOTICE", 5),
  ("be", 5),
  ("behalf", 5),
  ("but", 5),
  ("including", 5),
  ("license", 5),
  ("patent", 5),
  ("provided", 5),
  ("use", 5),
  ("work", 5),
  ("Apache", 4),
  ("CONDITIONS", 4),
  ("Object", 4),
  ("Work,", 4),
  ("additional", 4),
  ("attribution", 4),
  ("each", 4),
  ("form,", 4),
  ("if", 4),
  ("limited", 4),
  ("must", 4),
  ("only", 4),
  ("own", 4),
  ("owner", 4),
  ("part", 4),
  ("purposes", 4),
  ("required", 4),
  ("text", 4),
  ("use,", 4),
  ("whether", 4),
  ("your", 4),
  ("(including", 3),
  ("(or", 3),
  ("AND", 3),
  ("Contribution", 3),
  ("Contribution(s)", 3),
  ("Entity", 3),
  ("For", 3),
  ("Legal", 3),
  ("OF", 3),
  ("Unless", 3),
  ("Works;", 3),
  ("agreed", 3),
  ("all", 3),
  ("applicable", 3),
  ("by,", 3),
  ("communication", 3),
  ("copy", 3),
  ("damages", 3),
  ("distribute", 3),
  ("distributed", 3),
  ("distribution", 3),
  ("do", 3),
  ("entity", 3),
  ("excluding", 3),
  ("following", 3),
  ("from", 3),
  ("include", 3),
  ("individual", 3),
  ("its", 3),
  ("law", 3),
  ("modifications", 3),
  ("modifications,", 3),
  ("of,", 3),
  ("otherwise", 3),
  ("otherwise,", 3),
  ("permissions", 3),
  ("reproduction,", 3),
  ("those", 3),
  ("without", 3),
  ("writing,", 3),
  ("\"AS", 2),
  ("ANY", 2),
  ("BASIS,", 2),
  ("Contributions.", 2),
  ("Copyright", 2),
  ("FOR", 2),
  ("Grant", 2),
  ("IS\"", 2),
  ("If", 2),
  ("KIND,", 2),
  ("Liability.", 2),
  ("OR", 2),
  ("Subject", 2),
  ("TERMS", 2),
  ("The", 2),
  ("Version", 2),
  ("WARRANTIES", 2),
  ("WITHOUT", 2),
  ("Works,", 2),
  ("accepting", 2),
  ("add", 2),
  ("apply", 2),
  ("at", 2),
  ("authorized", 2),
  ("authorship,", 2),
  ("been", 2),
  ("cause", 2),
  ("claims", 2),
  ("code,", 2),
  ("control", 2),
  ("definition,", 2),
  ("direct", 2),
  ("distribute,", 2),
  ("documentation,", 2),
  ("either", 2),
  ("enclosed", 2),
  ("entity.", 2),
  ("except", 2),
  ("express", 2),
  ("generated", 2),
  ("granted", 2),
  ("grants", 2),
  ("has", 2),
  ("have", 2),
  ("hereby", 2),
  ("included", 2),
  ("inclusion", 2),
  ("incorporated", 2),
  ("indirect,", 2),
  ("intentionally", 2),
  ("irrevocable", 2),
  ("liability", 2),
  ("litigation", 2),
  ("means", 2),
  ("modify", 2),
  ("no", 2),
  ("no-charge,", 2),
  ("non-exclusive,", 2),
  ("notice", 2),
  ("original", 2),
  ("ownership", 2),
  ("perpetual,", 2),
  ("pertain", 2),
  ("provides", 2),
  ("publicly", 2),
  ("purpose", 2),
  ("redistributing", 2),
  ("royalty-free,", 2),
  ("sell,", 2),
  ("software", 2),
  ("source", 2),
  ("stated", 2),
  ("submitted", 2),
  ("their", 2),
  ("then", 2),
  ("thereof,", 2),
  ("third-party", 2),
  ("which", 2),
  ("whole,", 2),
  ("work,", 2),
  ("worldwide,", 2),
  ("you", 2),
  ("\"Contribution\"", 1),
  ("\"Contributor\"", 1),
  ("\"Derivative", 1),
  ("\"Legal", 1),
  ("\"License\"", 1),
  ("\"License\");", 1),
  ("\"Licensor\"", 1),
  ("\"NOTICE\"", 1),
  ("\"Not", 1),
  ("\"Object\"", 1),
  ("\"Source\"", 1),
  ("\"Work\"", 1),
  ("\"You\"", 1),
  ("\"Your\")", 1),
  ("\"[]\"", 1),
  ("\"control\"", 1),
  ("\"printed", 1),
  ("\"submitted\"", 1),
  ("(50%)", 1),
  ("(Don\'t", 1),
  ("(a)", 1),
  ("(an", 1),
  ("(and", 1),
  ("(b)", 1),
  ("(c)", 1),
  ("(d)", 1),
  ("(except", 1),
  ("(i)", 1),
  ("(ii)", 1),
  ("(iii)", 1),
  ("(such", 1),
  ("(the", 1),
  ("1", 1),
  ("1.", 1),
  ("2.", 1),
  ("2.0", 1),
  ("2.0,", 1),
  ("2004", 1),
  ("3.", 1),
  ("4.", 1),
  ("5.", 1),
  ("6.", 1),
  ("7.", 1),
  ("8.", 1),
  ("9", 1),
  ("9.", 1),
  ("A", 1),
  ("APPENDIX:", 1),
  ("Accepting", 1),
  ("Additional", 1),
  ("Appendix", 1),
  ("Contribution.\"", 1),
  ("Contributions)", 1),
  ("Contributor,", 1),
  ("DISTRIBUTION", 1),
  ("Definitions.", 1),
  ("Disclaimer", 1),
  ("END", 1),
  ("Entity\"", 1),
  ("FITNESS", 1),
  ("How", 1),
  ("However,", 1),
  ("In", 1),
  ("January", 1),
  ("License;", 1),
  ("Licensed", 1),
  ("Licensor,", 1),
  ("Limitation", 1),
  ("MERCHANTABILITY,", 1),
  ("NON-INFRINGEMENT,", 1),
  ("Notwithstanding", 1),
  ("PARTICULAR", 1),
  ("PURPOSE.", 1),
  ("Patent", 1),
  ("REPRODUCTION,", 1),
  ("Redistribution.", 1),
  ("Sections", 1),
  ("See", 1),
  ("Submission", 1),
  ("TITLE,", 1),
  ("This", 1),
  ("To", 1),
  ("Trademarks.", 1),
  ("USE,", 1),
  ("Warranty", 1),
  ("Warranty.", 1),
  ("We", 1),
  ("While", 1),
  ("Work.", 1),
  ("Works\"", 1),
  ("[name", 1),
  ("[yyyy]", 1),
  ("above,", 1),
  ("acceptance", 1),
  ("act", 1),
  ("acting", 1),
  ("acts)", 1),
  ("addendum", 1),
  ("additions", 1),
  ("advised", 1),
  ("against", 1),
  ("against,", 1),
  ("agree", 1),
  ("agreement", 1),
  ("alleging", 1),
  ("alone", 1),
  ("along", 1),
  ("alongside", 1),
  ("also", 1),
  ("and/or", 1),
  ("annotations,", 1),
  ("appear.", 1),
  ("applies", 1),
  ("appropriate", 1),
  ("appropriateness", 1),
  ("archives.", 1),
  ("arising", 1),
  ("asserted", 1),
  ("associated", 1),
  ("assume", 1),
  ("attach", 1),
  ("attached", 1),
  ("authorship.", 1),
  ("available", 1),
  ("based", 1),
  ("below).", 1),
  ("beneficial", 1),
  ("bind", 1),
  ("boilerplate", 1),
  ("brackets", 1),
  ("brackets!)", 1),
  ("cannot", 1),
  ("carry", 1),
  ("changed", 1),
  ("character", 1),
  ("charge", 1),
  ("choose", 1),
  ("class", 1),
  ("code", 1),
  ("combination", 1),
  ("comment", 1),
  ("commercial", 1),
  ("common", 1),
  ("compiled", 1),
  ("compliance", 1),
  ("complies", 1),
  ("computer", 1),
  ("conditions.", 1),
  ("conditions:", 1),
  ("configuration", 1),
  ("consequential", 1),
  ("consistent", 1),
  ("conspicuously", 1),
  ("constitutes", 1),
  ("construed", 1),
  ("contained", 1),
  ("content", 1),
  ("contents", 1),
  ("contract", 1),
  ("contract,", 1),
  ("contributory", 1),
  ("control,", 1),
  ("controlled", 1),
  ("conversions", 1),
  ("copies", 1),
  ("copyright,", 1),
  ("counterclaim", 1),
  ("cross-claim", 1),
  ("customary", 1),
  ("damages,", 1),
  ("damages.", 1),
  ("date", 1),
  ("defend,", 1),
  ("defined", 1),
  ("deliberate", 1),
  ("derived", 1),
  ("describing", 1),
  ("description", 1),
  ("designated", 1),
  ("determining", 1),
  ("different", 1),
  ("direct,", 1),
  ("direction", 1),
  ("discussing", 1),
  ("display", 1),
  ("display,", 1),
  ("distribution,", 1),
  ("document.", 1),
  ("documentation", 1),
  ("does", 1),
  ("easier", 1),
  ("editorial", 1),
  ("elaborations,", 1),
  ("electronic", 1),
  ("electronic,", 1),
  ("entities", 1),
  ("entity,", 1),
  ("even", 1),
  ("event", 1),
  ("example", 1),
  ("executed", 1),
  ("exercise", 1),
  ("exercising", 1),
  ("explicitly", 1),
  ("failure", 1),
  ("fee", 1),
  ("fields", 1),
  ("fifty", 1),
  ("file,", 1),
  ("file.", 1),
  ("filed.", 1),
  ("files", 1),
  ("files.", 1),
  ("files;", 1),
  ("for,", 1),
  ("form.", 1),
  ("format.", 1),
  ("from)", 1),
  ("from,", 1),
  ("give", 1),
  ("goodwill,", 1),
  ("governing", 1),
  ("grant", 1),
  ("granting", 1),
  ("grossly", 1),
  ("harmless", 1),
  ("herein", 1),
  ("hold", 1),
  ("http://www.apache.org/licenses/", 1),
  ("http://www.apache.org/licenses/LICENSE-2.0", 1),
  ("identification", 1),
  ("identifying", 1),
  ("implied,", 1),
  ("implied.", 1),
  ("import,", 1),
  ("improving", 1),
  ("inability", 1),
  ("incidental,", 1),
  ("includes", 1),
  ("including,", 1),
  ("incurred", 1),
  ("indemnify,", 1),
  ("indemnity,", 1),
  ("indicated", 1),
  ("information.", 1),
  ("informational", 1),
  ("infringed", 1),
  ("infringement,", 1),
  ("institute", 1),
  ("interfaces", 1),
  ("issue", 1),
  ("language", 1),
  ("lawsuit)", 1),
  ("least", 1),
  ("legal", 1),
  ("liability.", 1),
  ("liable", 1),
  ("licensable", 1),
  ("licenses", 1),
  ("limitation,", 1),
  ("limitations", 1),
  ("link", 1),
  ("lists,", 1),
  ("loss", 1),
  ("losses),", 1),
  ("made", 1),
  ("made,", 1),
  ("mailing", 1),
  ("make,", 1),
  ("making", 1),
  ("malfunction,", 1),
  ("managed", 1),
  ("management", 1),
  ("marked", 1),
  ("marks,", 1),
  ("mechanical", 1),
  ("media", 1),
  ("medium,", 1),
  ("meet", 1),
  ("merely", 1),
  ("modified", 1),
  ("modifying", 1),
  ("more", 1),
  ("name", 1),
  ("name)", 1),
  ("names", 1),
  ("names,", 1),
  ("necessarily", 1),
  ("negligence),", 1),
  ("negligent", 1),
  ("normally", 1),
  ("nothing", 1),
  ("notice,", 1),
  ("object", 1),
  ("obligations", 1),
  ("obligations,", 1),
  ("obtain", 1),
  ("offer", 1),
  ("offer,", 1),
  ("one", 1),
  ("or,", 1),
  ("origin", 1),
  ("out", 1),
  ("outstanding", 1),
  ("owner.", 1),
  ("owner]", 1),
  ("page\"", 1),
  ("patent,", 1),
  ("percent", 1),
  ("perform,", 1),
  ("permission", 1),
  ("places:", 1),
  ("possibility", 1),
  ("power,", 1),
  ("preferred", 1),
  ("prepare", 1),
  ("product", 1),
  ("prominent", 1),
  ("provide", 1),
  ("readable", 1),
  ("reason", 1),
  ("reasonable", 1),
  ("received", 1),
  ("recipients", 1),
  ("recommend", 1),
  ("regarding", 1),
  ("remain", 1),
  ("replaced", 1),
  ("represent,", 1),
  ("representatives,", 1),
  ("reproduce", 1),
  ("reproduce,", 1),
  ("reproducing", 1),
  ("responsibility,", 1),
  ("responsible", 1),
  ("result", 1),
  ("resulting", 1),
  ("retain,", 1),
  ("revisions,", 1),
  ("rights", 1),
  ("risks", 1),
  ("same", 1),
  ("section)", 1),
  ("sent", 1),
  ("separable", 1),
  ("separate", 1),
  ("service", 1),
  ("shares,", 1),
  ("should", 1),
  ("sole", 1),
  ("solely", 1),
  ("source,", 1),
  ("special,", 1),
  ("specific", 1),
  ("state", 1),
  ("statement", 1),
  ("stating", 1),
  ("stoppage,", 1),
  ("sublicense,", 1),
  ("submit", 1),
  ("submitted.", 1),
  ("subsequently", 1),
  ("supersede", 1),
  ("support,", 1),
  ("syntax", 1),
  ("systems", 1),
  ("systems,", 1),
  ("terminate", 1),
  ("theory,", 1),
  ("thereof", 1),
  ("thereof.", 1),
  ("through", 1),
  ("tort", 1),
  ("tracking", 1),
  ("trade", 1),
  ("trademark,", 1),
  ("trademarks,", 1),
  ("transfer", 1),
  ("transformation", 1),
  ("translation", 1),
  ("types.", 1),
  ("union", 1),
  ("unless", 1),
  ("using", 1),
  ("verbal,", 1),
  ("version", 1),
  ("warranties", 1),
  ("warranty", 1),
  ("warranty,", 1),
  ("was", 1),
  ("where", 1),
  ("wherever", 1),
  ("whom", 1),
  ("work.", 1),
  ("works", 1),
  ("writing", 1),
  ("written", 1),
]
